akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala

```diff
@@ -121,17 +144,26 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with StageLoggi
     protected def maybeFinishStream(endStream: Boolean): IncomingStreamState =
       if (endStream) afterEndStreamReceived else this
   }
+
   // on the incoming side there's (almost) no difference between Open and HalfClosedLocal
-  case class Open(buffer: IncomingStreamBuffer) extends ReceivingData(HalfClosedRemote)
-  // currently unused: we never close a stream before the peer does
-  // case class HalfClosedLocal(buffer: Buffer) extends ReceivingData(Closed)
+  case class Open(buffer: IncomingStreamBuffer) extends ReceivingData(HalfClosedRemote) {
+    override def handleOutgoingEnded(): IncomingStreamState = HalfClosedLocal(buffer)
+  }
+  case class HalfClosedLocal(buffer: IncomingStreamBuffer) extends ReceivingData(Closed) {
+    override def handleOutgoingEnded(): IncomingStreamState = {
```

Comment:
Maybe it could be the default implementation in `IncomingStreamState` with a message "handleOutgoingEnded received unexpectedly in state $productPrefix. This indicates a bug in Akka HTTP, please report it to the issue tracker."?

Reply:
It's a trade-off: while this is more verbose, it also forces us to consider whether the signal that the outgoing side was closed should have an effect for each state separately. Given the state diagram should be mostly static I agree we can use a default implementation, though - will update.